### PR TITLE
#328: Query for `getLocalMaxTimestamp`

### LINKED ISSUE #328: Reduce number of queries dispatched by `getLocalMaxTimestamp`
One of the first queries executed by sync is [`getLocalMaxTimestamp`](https://github.com/alephium/explorer-backend/blob/master/app/src/main/scala/org/alephium/explorer/service/BlockFlowSyncService.scala#L181) to calculate maximum `height` and `timestamp`. Both these fields are in `block_headers` table. 

But the query dispatches a number of other queries accessing multiple tables `transactions`, `block_deps`, `outputs`, `inputs` which are not needed other than just `block_headers`

Incase I'm overlooking something. Is this a valid issue? @tdroxler 

## Example un-tested query

Query to maybe use instead (un-tested)

```sql
SELECT (max(block_timestamp),
        sum(height),
        height)
FROM block_headers
WHERE chain_from = 1
  AND chain_to = 4
  AND height = (SELECT max(height)
                from block_headers
                WHERE chain_from = 1
                  AND chain_to = 4)
group by height;
```

## Use `UNION`

[The function](https://github.com/alephium/explorer-backend/blob/master/app/src/main/scala/org/alephium/explorer/service/BlockFlowSyncService.scala#L181) also traverses over multiple groups and queries them concurrently which can be replaced with `UNION`.

### PR DESCRIPTION
**Pull request by simerplaha**
- Resolves #328. 
- Ran the test-case in this PR over 100s of iterations. The test also checks for same output as existing code. So it should cover all existing cases. 

Started off with typed query [here](https://github.com/alephium/explorer-backend/blob/328_getLocalMaxTimestamp_2/app/src/main/scala/org/alephium/explorer/persistence/queries/BlockQueries.scala#L342) hoping to reuse existing queries to speed things up. But the generated `UNION` queries looked like following so switched to SQL. 

## TODO
- Remove deprecated code. See TODO in code. 
- Benchmarks. Please let me if you'd like this done.

## Generated query
<details>
  <summary>Generated query</summary>

  ```sql
(select x2.x3 as x4, x2.x5 as x6
 from ((select x7.x8 as x3, x7.x9 as x5
        from ((select x10.x11 as x8, x10.x12 as x9
               from ((select x13.x14 as x11, x13.x15 as x12
                      from ((select x16.x17 as x14, x16.x18 as x15
                             from ((select x19.x20 as x17, x19.x21 as x18
                                    from ((select x22.x23 as x20, x22.x24 as x21
                                           from ((select "height" as x23, max("block_timestamp") as x24
                                                  from "block_headers"
                                                  where (("height" = (select max("height")
                                                                      from "block_headers"
                                                                      where ("chain_from" = 0)
                                                                        and ("chain_to" = 0))) and ("chain_from" = 0))
                                                    and ("chain_to" = 0)
                                                  group by "height")
                                                 union all
                                                 (select "height" as x23, max("block_timestamp") as x24
                                                  from "block_headers"
                                                  where (("height" = (select max("height")
                                                                      from "block_headers"
                                                                      where ("chain_from" = 0)
                                                                        and ("chain_to" = 1))) and ("chain_from" = 0))
                                                    and ("chain_to" = 1)
                                                  group by "height")) x22)
                                          union all
                                          (select "height" as x20, max("block_timestamp") as x21
                                           from "block_headers"
                                           where (("height" = (select max("height")
                                                               from "block_headers"
                                                               where ("chain_from" = 0) and ("chain_to" = 2))) and
                                                  ("chain_from" = 0))
                                             and ("chain_to" = 2)
                                           group by "height")) x19)
                                   union all
                                   (select "height" as x17, max("block_timestamp") as x18
                                    from "block_headers"
                                    where (("height" = (select max("height")
                                                        from "block_headers"
                                                        where ("chain_from" = 1) and ("chain_to" = 0))) and
                                           ("chain_from" = 1))
                                      and ("chain_to" = 0)
                                    group by "height")) x16)
                            union all
                            (select "height" as x14, max("block_timestamp") as x15
                             from "block_headers"
                             where (("height" = (select max("height")
                                                 from "block_headers"
                                                 where ("chain_from" = 1) and ("chain_to" = 1))) and ("chain_from" = 1))
                               and ("chain_to" = 1)
                             group by "height")) x13)
                     union all
                     (select "height" as x11, max("block_timestamp") as x12
                      from "block_headers"
                      where (("height" = (select max("height")
                                          from "block_headers"
                                          where ("chain_from" = 1) and ("chain_to" = 2))) and ("chain_from" = 1))
                        and ("chain_to" = 2)
                      group by "height")) x10)
              union all
              (select "height" as x8, max("block_timestamp") as x9
               from "block_headers"
               where (("height" =
                       (select max("height") from "block_headers" where ("chain_from" = 2) and ("chain_to" = 0))) and
                      ("chain_from" = 2))
                 and ("chain_to" = 0)
               group by "height")) x7)
       union all
       (select "height" as x3, max("block_timestamp") as x5
        from "block_headers"
        where (("height" =
                (select max("height") from "block_headers" where ("chain_from" = 2) and ("chain_to" = 1))) and
               ("chain_from" = 2))
          and ("chain_to" = 1)
        group by "height")) x2)
union all
(select "height" as x4, max("block_timestamp") as x6
 from "block_headers"
 where (("height" = (select max("height") from "block_headers" where ("chain_from" = 2) and ("chain_to" = 2))) and
        ("chain_from" = 2))
   and ("chain_to" = 2)
 group by "height");
  ```
</details>

### CODE REVIEW

File: app/src/main/scala/org/alephium/explorer/persistence/queries/BlockQueries.scala [hunk 1]
```diff
@@ -311,4 +313,63 @@ object BlockQueries extends StrictLogging {
...
+
+  /**
+    * Fetches the maximum `block_timestamp` and sum height of
+    * all blocks filtered with maximum height for all input chain
+    * chains provided by [[GroupSetting.groupIndexes]]
+    *
+    * @param groupSetting Provides the list of chains to run this query on
+    * @return Maximum timestamp and sum of all heights i.e. the total number of blocks.
+    */
+  def noOfBlocksAndMaxBlockTimestamp()(
```
Comment (tdroxler) [inline app/src/main/scala/org/alephium/explorer/persistence/queries/BlockQueries.scala:343]: super minor: I spent few seconds understanding why the function start with `no` (likethe contrary of `yes`.... ), maybe we could use `numOfBlocksAndMaxBlockTimestamp` ?

File: app/src/main/scala/org/alephium/explorer/persistence/model/BlockHeader.scala [hunk 2]
```diff
@@ -65,4 +67,62 @@ object BlockHeader {
...
+
+  /** Sum height of all blocks */
+  def sumHeight(blocks: Iterable[BlockHeader]): Option[Height] =
+    blocks.foldLeft(Option.empty[Height]) {
+      case (currentSum, header) =>
+        currentSum match {
+          case Some(sum) => Some(Height.unsafe(sum.value + header.height.value))
+          case None      => Some(header.height)
+        }
+    }
```
Comment (tdroxler) [inline app/src/main/scala/org/alephium/explorer/persistence/model/BlockHeader.scala:127]: all above functions are only used in tests, I would rather move them there

**Review by tdroxler — APPROVED** [review]
Minor comments, but LGTM :rocket: 
As discussed privately, we can maybe simplify a lot the logic I wrote for that part of the code, but in the mean time your code is much more better, nice job.

File: app/src/main/scala/org/alephium/explorer/persistence/model/BlockHeader.scala (hunk 2 above)
Comment (simerplaha) [reply app/src/main/scala/org/alephium/explorer/persistence/model/BlockHeader.scala]: I wasn't sure where in test you'd preferred them to be. I've placed them in `BlockHeaderSpec`. Let me know if you'd like them moved elsewhere.

**Pull request merged**